File: oop-fundamentals/6.inheritance/ts/inheritance.ts
//
// =======================================================
// 1. WHAT IS INHERITANCE?
// =======================================================
//
// Inheritance allows one class (child/subclass) to inherit properties
// and behaviors from another class (parent/superclass).
//
// Key formula:
// Inheritance = Code Reuse + "is-a" Relationship + Hierarchy
//
// Benefits:
// - Code reusability (write once, use many)
// - Logical hierarchy (models real-world relationships)
// - Ease of maintenance (change once, affects all)
// - Enables polymorphism (runtime behavior)
//
// Mental Model:
// Think of a family tree: children inherit traits from parents,
// but can also have unique characteristics.

//
// =======================================================
// 2. BASIC SYNTAX
// =======================================================
//
// TypeScript uses 'extends' keyword for inheritance
// Single inheritance only (no multiple inheritance like C++)

class Animal {
  protected name: string;
  protected age: number;

  constructor(name: string, age: number) {
    this.name = name;
    this.age = age;
    console.log(`🐾 Animal constructor: ${name}`);
  }

  makeSound(): void {
    console.log(`${this.name} makes a sound`);
  }

  sleep(): void {
    console.log(`${this.name} is sleeping 💤`);
  }

  displayInfo(): void {
    console.log(`Name: ${this.name} | Age: ${this.age}`);
  }
}

class Dog extends Animal {
  private breed: string;

  constructor(name: string, age: number, breed: string) {
    super(name, age); // Must call super() first!
    this.breed = breed;
    console.log(`🐕 Dog constructor: ${name}`);
  }

  // Override parent method
  makeSound(): void {
    console.log(`${this.name} barks: Woof! Woof! 🐕`);
  }

  // Dog-specific method
  fetch(): void {
    console.log(`${this.name} is fetching the ball! 🎾`);
  }

  displayBreed(): void {
    console.log(`${this.name} is a ${this.breed}`);
  }
}

//
// =======================================================
// 3. TYPES OF INHERITANCE
// =======================================================
//
// TypeScript supports:
// 1. Single Inheritance (one parent only)
// 2. Multi-level Inheritance (chain)
// 3. Hierarchical Inheritance (multiple children)
//
// Does NOT support:
// - Multiple Inheritance (use interfaces instead)

// ----- Single Inheritance (shown above) -----

// ----- Multi-level Inheritance -----
class Vehicle {
  protected make: string;
  protected year: number;

  constructor(make: string, year: number) {
    this.make = make;
    this.year = year;
  }

  start(): void {
    console.log("🔑 Vehicle starting...");
  }

  displayInfo(): void {
    console.log(`${this.year} ${this.make}`);
  }
}

class Car extends Vehicle {
  protected numDoors: number;

  constructor(make: string, year: number, numDoors: number) {
    super(make, year);
    this.numDoors = numDoors;
  }

  start(): void {
    console.log("🚗 Car engine starting...");
  }

  honk(): void {
    console.log("🎺 Beep beep!");
  }
}

// Multi-level: ElectricCar extends Car, which extends Vehicle
class ElectricCar extends Car {
  private batteryCapacity: number;

  constructor(make: string, year: number, numDoors: number, battery: number) {
    super(make, year, numDoors);
    this.batteryCapacity = battery;
  }

  start(): void {
    console.log("⚡ Electric car starting silently...");
  }

  charge(): void {
    console.log(`🔌 Charging ${this.batteryCapacity} kWh battery`);
  }
}

// ----- Hierarchical Inheritance -----
abstract class Shape {
  protected color: string;

  constructor(color: string) {
    this.color = color;
  }

  abstract getArea(): number;

  display(): void {
    console.log(`Color: ${this.color}`);
  }
}

class Circle extends Shape {
  private radius: number;

  constructor(color: string, radius: number) {
    super(color);
    this.radius = radius;
  }

  getArea(): number {
    return Math.PI * this.radius ** 2;
  }

  display(): void {
    console.log(
      `⭕ Circle | Color: ${this.color} | Area: ${this.getArea().toFixed(2)}`,
    );
  }
}

class Rectangle extends Shape {
  private width: number;
  private height: number;

  constructor(color: string, width: number, height: number) {
    super(color);
    this.width = width;
    this.height = height;
  }

  getArea(): number {
    return this.width * this.height;
  }

  display(): void {
    console.log(`▭ Rectangle | Color: ${this.color} | Area: ${this.getArea()}`);
  }
}

//
// =======================================================
// 4. MULTIPLE INHERITANCE (via INTERFACES)
// =======================================================
//
// TypeScript doesn't support multiple class inheritance
// But you CAN implement multiple interfaces!

interface Flyable {
  fly(): void;
}

interface Swimmable {
  swim(): void;
}

class Duck extends Animal implements Flyable, Swimmable {
  constructor(name: string, age: number) {
    super(name, age);
  }

  makeSound(): void {
    console.log(`${this.name} quacks: Quack! Quack! 🦆`);
  }

  fly(): void {
    console.log(`${this.name} is flying! ✈️`);
  }

  swim(): void {
    console.log(`${this.name} is swimming! 🏊`);
  }
}

//
// =======================================================
// 5. SUPER KEYWORD
// =======================================================
//
// 'super' is used to:
// 1. Call parent constructor (must be first in child constructor)
// 2. Call parent methods

class Employee {
  protected name: string;
  protected salary: number;

  constructor(name: string, salary: number) {
    this.name = name;
    this.salary = salary;
  }

  work(): void {
    console.log(`${this.name} is working...`);
  }

  getInfo(): string {
    return `${this.name} earns $${this.salary}`;
  }
}

class Manager extends Employee {
  private teamSize: number;

  constructor(name: string, salary: number, teamSize: number) {
    super(name, salary); // Call parent constructor
    this.teamSize = teamSize;
  }

  work(): void {
    super.work(); // Call parent method
    console.log(`${this.name} is managing ${this.teamSize} people`);
  }

  getInfo(): string {
    return super.getInfo() + ` and manages ${this.teamSize} people`;
  }
}

//
// =======================================================
// 6. ACCESS MODIFIERS IN INHERITANCE
// =======================================================
//
// - public: Accessible everywhere
// - protected: Accessible in class and subclasses
// - private: Only in the class itself
// - readonly: Can't modify after initialization

class BankAccount {
  public accountNumber: string;
  protected balance: number;
  private pin: string;
  readonly createdAt: Date;

  constructor(accountNumber: string, pin: string) {
    this.accountNumber = accountNumber;
    this.balance = 0;
    this.pin = pin;
    this.createdAt = new Date();
  }

  protected validatePin(inputPin: string): boolean {
    return this.pin === inputPin;
  }

  getBalance(): number {
    return this.balance;
  }
}

class SavingsAccount extends BankAccount {
  private interestRate: number;

  constructor(accountNumber: string, pin: string, rate: number) {
    super(accountNumber, pin);
    this.interestRate = rate;
  }

  addInterest(): void {
    // Can access protected 'balance'
    const interest = this.balance * this.interestRate;
    this.balance += interest;
    console.log(`✅ Added interest: $${interest.toFixed(2)}`);
  }

  // Can access protected method
  withdraw(amount: number, pin: string): void {
    if (!this.validatePin(pin)) {
      console.log("❌ Invalid PIN");
      return;
    }
    if (amount > this.balance) {
      console.log("❌ Insufficient funds");
      return;
    }
    this.balance -= amount;
    console.log(`✅ Withdrew $${amount}`);
  }

  // Cannot access private 'pin' directly!
}

//
// =======================================================
// 7. REAL-WORLD EXAMPLE: NOTIFICATION SYSTEM
// =======================================================
//

class Notification {
  protected recipient: string;
  protected message: string;
  protected timestamp: Date;

  constructor(recipient: string, message: string) {
    this.recipient = recipient;
    this.message = message;
    this.timestamp = new Date();
  }

  formatHeader(): string {
    return `[${this.timestamp.toISOString()}] To: ${this.recipient}`;
  }

  send(): void {
    console.log(this.formatHeader());
    console.log(`Message: ${this.message}`);
  }
}

class EmailNotification extends Notification {
  private subject: string;

  constructor(recipient: string, message: string, subject: string) {
    super(recipient, message);
    this.subject = subject;
  }

  send(): void {
    console.log("\n📧 EMAIL NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Subject: ${this.subject}`);
    console.log(`Body: ${this.message}`);
    console.log("✅ Email delivered");
  }
}

class SMSNotification extends Notification {
  private phoneNumber: string;
  private static readonly MAX_LENGTH = 160;

  constructor(recipient: string, message: string, phoneNumber: string) {
    super(recipient, message);
    this.phoneNumber = phoneNumber;
  }

  send(): void {
    console.log("\n📱 SMS NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Phone: ${this.phoneNumber}`);

    const smsBody =
      this.message.length > SMSNotification.MAX_LENGTH
        ? this.message.substring(0, SMSNotification.MAX_LENGTH - 3) + "..."
        : this.message;

    console.log(`SMS: ${smsBody}`);
    console.log(
      `✅ SMS sent (${smsBody.length}/${SMSNotification.MAX_LENGTH} chars)`,
    );
  }
}

class PushNotification extends Notification {
  private deviceToken: string;
  private priority: string;

  constructor(
    recipient: string,
    message: string,
    deviceToken: string,
    priority: string,
  ) {
    super(recipient, message);
    this.deviceToken = deviceToken;
    this.priority = priority;
  }

  send(): void {
    console.log("\n🔔 PUSH NOTIFICATION");
    console.log(this.formatHeader());
    console.log(`Device: ${this.deviceToken.substring(0, 8)}...`);
    console.log(`Priority: ${this.priority}`);
    console.log(`Alert: ${this.message}`);
    console.log("✅ Push delivered");
  }
}

//
// =======================================================
// 8. POLYMORPHISM THROUGH INHERITANCE
// =======================================================
//

function sendNotification(notification: Notification): void {
  notification.send(); // Polymorphic call!
}

//
// =======================================================
// 9. METHOD OVERRIDING
// =======================================================
//

class Payment {
  protected amount: number;

  constructor(amount: number) {
    this.amount = amount;
  }

  process(): void {
    console.log(`Processing payment of $${this.amount}`);
  }

  getType(): string {
    return "Generic Payment";
  }
}

class CreditCardPayment extends Payment {
  private cardNumber: string;

  constructor(amount: number, cardNumber: string) {
    super(amount);
    this.cardNumber = cardNumber;
  }

  // Override parent method
  process(): void {
    console.log(`💳 Processing credit card payment of $${this.amount}`);
    console.log(`Card: ${this.cardNumber.substring(0, 4)}****`);
  }

  getType(): string {
    return "Credit Card";
  }
}

class PayPalPayment extends Payment {
  private email: string;

  constructor(amount: number, email: string) {
    super(amount);
    this.email = email;
  }

  process(): void {
    console.log(`💰 Processing PayPal payment of $${this.amount}`);
    console.log(`Account: ${this.email}`);
  }

  getType(): string {
    return "PayPal";
  }
}

//
// =======================================================
// 10. ABSTRACT CLASSES
// =======================================================
//
// Abstract classes can't be instantiated
// They serve as base classes for other classes

abstract class DatabaseConnection {
  protected connectionString: string;
  protected isConnected: boolean = false;

  constructor(connectionString: string) {
    this.connectionString = connectionString;
  }

  // Abstract method - must be implemented by subclasses
  abstract connect(): void;
  abstract disconnect(): void;
  abstract executeQuery(query: string): void;

  // Concrete method - can be used by all subclasses
  getStatus(): string {
    return this.isConnected ? "Connected" : "Disconnected";
  }
}

class MySQLConnection extends DatabaseConnection {
  connect(): void {
    console.log(`🔗 Connecting to MySQL at ${this.connectionString}`);
    this.isConnected = true;
  }

  disconnect(): void {
    console.log("🔌 Disconnecting from MySQL");
    this.isConnected = false;
  }

  executeQuery(query: string): void {
    if (!this.isConnected) {
      console.log("❌ Not connected");
      return;
    }
    console.log(`📊 Executing MySQL query: ${query}`);
  }
}

class MongoDBConnection extends DatabaseConnection {
  connect(): void {
    console.log(`🔗 Connecting to MongoDB at ${this.connectionString}`);
    this.isConnected = true;
  }

  disconnect(): void {
    console.log("🔌 Disconnecting from MongoDB");
    this.isConnected = false;
  }

  executeQuery(query: string): void {
    if (!this.isConnected) {
      console.log("❌ Not connected");
      return;
    }
    console.log(`📊 Executing MongoDB query: ${query}`);
  }
}

//
// =======================================================
// 11. DEMONSTRATION
// =======================================================
//

function main(): void {
  console.log("=== Inheritance in TypeScript Demo ===\n");

  // ---- Basic Inheritance ----
  console.log("1. Basic Inheritance (Animal → Dog):");
  const dog = new Dog("Buddy", 3, "Golden Retriever");
  dog.makeSound();
  dog.sleep();
  dog.fetch();
  dog.displayBreed();

  // ---- Multi-level Inheritance ----
  console.log("\n2. Multi-level Inheritance (Vehicle → Car → ElectricCar):");
  const tesla = new ElectricCar("Tesla", 2024, 4, 100);
  tesla.start();
  tesla.charge();
  tesla.honk();

  // ---- Hierarchical Inheritance ----
  console.log("\n3. Hierarchical Inheritance (Shape → Circle, Rectangle):");
  const circle = new Circle("Red", 5);
  const rectangle = new Rectangle("Blue", 10, 20);

  circle.display();
  rectangle.display();

  // ---- Multiple Interfaces ----
  console.log("\n4. Multiple Interfaces (Duck: Animal + Flyable + Swimmable):");
  const duck = new Duck("Donald", 2);
  duck.makeSound();
  duck.fly();
  duck.swim();

  // ---- Super Keyword ----
  console.log("\n5. Super Keyword (Employee → Manager):");
  const manager = new Manager("Alice", 120000, 5);
  manager.work();
  console.log(manager.getInfo());

  // ---- Access Modifiers ----
  console.log("\n6. Access Modifiers (BankAccount → SavingsAccount):");
  const savings = new SavingsAccount("SA-001", "1234", 0.05);
  savings["balance"] = 1000; // Set balance (normally via deposit method)
  savings.addInterest();
  savings.withdraw(100, "1234");

  // ---- Notification System ----
  console.log("\n7. Real-World Example (Notification System):");
  const email = new EmailNotification(
    "alice@example.com",
    "Your order has been shipped!",
    "Order Update",
  );
  email.send();

  const sms = new SMSNotification(
    "Bob",
    "Your verification code is 482910.",
    "+1-555-0123",
  );
  sms.send();

  const push = new PushNotification(
    "Charlie",
    "New message from Alice",
    "d8a3f4b2c1e5a9b7",
    "high",
  );
  push.send();

  // ---- Polymorphism ----
  console.log("\n8. Polymorphism Through Inheritance:");
  const notifications: Notification[] = [email, sms, push];

  notifications.forEach((notification) => {
    sendNotification(notification);
  });

  // ---- Method Overriding ----
  console.log("\n9. Method Overriding (Payment System):");
  const creditCard = new CreditCardPayment(100, "1234567890123456");
  const paypal = new PayPalPayment(50, "user@example.com");

  const payments: Payment[] = [creditCard, paypal];
  payments.forEach((payment) => {
    console.log(`\nPayment Type: ${payment.getType()}`);
    payment.process();
  });

  // ---- Abstract Classes ----
  console.log("\n10. Abstract Classes (Database Connections):");
  const mysql: DatabaseConnection = new MySQLConnection("localhost:3306");
  mysql.connect();
  mysql.executeQuery("SELECT * FROM users");
  console.log(`Status: ${mysql.getStatus()}`);
  mysql.disconnect();

  console.log();

  const mongo: DatabaseConnection = new MongoDBConnection("localhost:27017");
  mongo.connect();
  mongo.executeQuery("db.users.find()");
  mongo.disconnect();

  console.log("\n=== Demo Complete ===");
}

// Execute
main();

export {};

/*
📘 Key Insights:

1. **Inheritance in TypeScript**:
   - Single inheritance only (extends one class)
   - Can implement multiple interfaces
   - No multiple class inheritance (unlike C++)

2. **Super Keyword**:
   - Must call super() in constructor (first statement!)
   - Use super.method() to call parent methods
   - Enables method extension, not just replacement

3. **Access Modifiers**:
   - public: Accessible everywhere (default)
   - protected: Accessible in class and subclasses
   - private: Only in the class itself
   - # private: True runtime privacy (ES2022+)

4. **Method Overriding**:
   - Override parent methods without special keyword
   - Use super to call parent version
   - Enables polymorphic behavior

5. **Abstract Classes**:
   - Can't instantiate directly
   - abstract methods must be implemented
   - Can have concrete methods
   - Good for defining contracts with shared logic

Best Practices:
- Always call super() first in constructor
- Use protected for inherited members
- Prefer composition over inheritance when possible
- Keep inheritance hierarchies shallow
- Use interfaces for multiple inheritance of behavior
- Use abstract classes for shared behavior

When to Use Inheritance:
✅ Clear "is-a" relationship
✅ Shared behavior across classes
✅ Need polymorphism
✅ Logical hierarchy exists

When to Avoid:
❌ "has-a" relationship (use composition)
❌ Need multiple inheritance (use interfaces)
❌ Deep hierarchies (>3 levels)
❌ Runtime flexibility needed
*/

File: 1.oop-fundamentals/6.inheritance/cpp/inheritance.cpp
#include <ctime>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

// Ref - https://algomaster.io/learn/lld/inheritance
// Ref - https://www.geeksforgeeks.org/inheritance-in-c/
// Ref - https://en.cppreference.com/w/cpp/language/derived_class

//
// =======================================================
// 1. WHAT IS INHERITANCE?
// =======================================================
//
// Inheritance allows one class (subclass/child) to inherit properties
// and behaviors of another class (superclass/parent).
//
// Key formula:
// Inheritance = Code Reuse + "is-a" Relationship + Hierarchy
//
// Benefits:
// - Code reusability (write once, use many)
// - Logical hierarchy (models real-world relationships)
// - Ease of maintenance (change once, affects all)
// - Enables polymorphism (runtime behavior)
//
// Mental Model:
// Think of a family tree: children inherit traits from parents,
// but can also have unique characteristics.

//
// =======================================================
// 2. BASIC SYNTAX & ACCESS SPECIFIERS
// =======================================================
//
// C++ supports three inheritance modes: public, protected, private
//
// class Derived : <access_specifier> Base { };
//
// Public inheritance (most common):
// - public members → public in derived
// - protected members → protected in derived
// - private members → not accessible
//
// Protected inheritance (rare):
// - public members → protected in derived
// - protected members → protected in derived
//
// Private inheritance (very rare):
// - public members → private in derived
// - protected members → private in derived

class Animal {
protected:
  std::string name;
  int age;

public:
  Animal(const std::string &n, int a) : name(n), age(a) {
    std::cout << "🐾 Animal constructor: " << name << "\n";
  }

  virtual ~Animal() { std::cout << "🐾 Animal destructor: " << name << "\n"; }

  virtual void makeSound() const { std::cout << name << " makes a sound\n"; }

  void sleep() const { std::cout << name << " is sleeping 💤\n"; }

  void displayInfo() const {
    std::cout << "Name: " << name << " | Age: " << age << "\n";
  }
};

class Dog : public Animal {
private:
  std::string breed;

public:
  Dog(const std::string &n, int a, const std::string &b)
      : Animal(n, a), breed(b) {
    std::cout << "🐕 Dog constructor: " << name << "\n";
  }

  ~Dog() { std::cout << "🐕 Dog destructor: " << name << "\n"; }

  // Override parent method
  void makeSound() const override {
    std::cout << name << " barks: Woof! Woof! 🐕\n";
  }

  // Dog-specific method
  void fetch() const { std::cout << name << " is fetching the ball! 🎾\n"; }

  void displayBreed() const { std::cout << name << " is a " << breed << "\n"; }
};

//
// =======================================================
// 3. TYPES OF INHERITANCE
// =======================================================
//
// 1. Single Inheritance: One child, one parent
// 2. Multi-level Inheritance: Chain of inheritance
// 3. Hierarchical Inheritance: Multiple children, one parent
// 4. Multiple Inheritance: One child, multiple parents (C++ only)

// ----- Single Inheritance (already shown above) -----

// ----- Multi-level Inheritance -----
class Vehicle {
protected:
  std::string make;
  int year;

public:
  Vehicle(const std::string &m, int y) : make(m), year(y) {}

  virtual void start() { std::cout << "🔑 Vehicle starting...\n"; }

  void displayInfo() const { std::cout << year << " " << make << "\n"; }
};

class Car : public Vehicle {
protected:
  int numDoors;

public:
  Car(const std::string &m, int y, int doors)
      : Vehicle(m, y), numDoors(doors) {}

  void start() override { std::cout << "🚗 Car engine starting...\n"; }

  void honk() const { std::cout << "🎺 Beep beep!\n"; }
};

// Multi-level: ElectricCar inherits from Car, which inherits from Vehicle
class ElectricCar : public Car {
private:
  int batteryCapacity;

public:
  ElectricCar(const std::string &m, int y, int doors, int battery)
      : Car(m, y, doors), batteryCapacity(battery) {}

  void start() override {
    std::cout << "⚡ Electric car starting silently...\n";
  }

  void charge() const {
    std::cout << "🔌 Charging " << batteryCapacity << " kWh battery\n";
  }
};

// ----- Hierarchical Inheritance -----
// Multiple classes inherit from same base

class Shape {
protected:
  std::string color;

public:
  Shape(const std::string &c) : color(c) {}

  virtual double getArea() const = 0; // Pure virtual

  virtual void display() const { std::cout << "Color: " << color << "\n"; }

  virtual ~Shape() = default;
};

class Circle : public Shape {
private:
  double radius;

public:
  Circle(const std::string &c, double r) : Shape(c), radius(r) {}

  double getArea() const override { return 3.14159 * radius * radius; }

  void display() const override {
    std::cout << "⭕ Circle | ";
    Shape::display();
    std::cout << "Area: " << getArea() << "\n";
  }
};

class Rectangle : public Shape {
private:
  double width, height;

public:
  Rectangle(const std::string &c, double w, double h)
      : Shape(c), width(w), height(h) {}

  double getArea() const override { return width * height; }

  void display() const override {
    std::cout << "▭ Rectangle | ";
    Shape::display();
    std::cout << "Area: " << getArea() << "\n";
  }
};

//
// =======================================================
// 4. MULTIPLE INHERITANCE
// =======================================================
//
// C++ allows a class to inherit from multiple base classes
// This can lead to the "Diamond Problem"

class Flyable {
public:
  virtual void fly() const { std::cout << "✈️  Flying in the air!\n"; }
  virtual ~Flyable() = default;
};

class Swimmable {
public:
  virtual void swim() const { std::cout << "🏊 Swimming in water!\n"; }
  virtual ~Swimmable() = default;
};

// Multiple inheritance: Duck can both fly and swim
class Duck : public Animal, public Flyable, public Swimmable {
public:
  Duck(const std::string &name, int age) : Animal(name, age) {}

  void makeSound() const override {
    std::cout << name << " quacks: Quack! Quack! 🦆\n";
  }
};

//
// =======================================================
// 5. THE DIAMOND PROBLEM & VIRTUAL INHERITANCE
// =======================================================
//
// Diamond Problem occurs when:
// - Class D inherits from B and C
// - Both B and C inherit from A
// - D gets two copies of A's members!
//
// Solution: Virtual inheritance

class PowerSource {
protected:
  int powerLevel;

public:
  PowerSource(int level = 100) : powerLevel(level) {
    std::cout << "⚡ PowerSource created (Level: " << powerLevel << ")\n";
  }

  void displayPower() const {
    std::cout << "Power Level: " << powerLevel << "%\n";
  }
};

// Virtual inheritance prevents duplicate PowerSource
class ElectricComponent : virtual public PowerSource {
public:
  ElectricComponent(int level = 100) : PowerSource(level) {
    std::cout << "🔌 ElectricComponent initialized\n";
  }
};

class BatteryComponent : virtual public PowerSource {
public:
  BatteryComponent(int level = 100) : PowerSource(level) {
    std::cout << "🔋 BatteryComponent initialized\n";
  }
};

// Without virtual: would have 2 PowerSource instances
// With virtual: only 1 PowerSource instance
class HybridDevice : public ElectricComponent, public BatteryComponent {
public:
  HybridDevice(int level = 100)
      : PowerSource(level), // Must initialize virtual base explicitly
        ElectricComponent(level), BatteryComponent(level) {
    std::cout << "🤖 HybridDevice ready\n";
  }
};

//
// =======================================================
// 6. CONSTRUCTOR & DESTRUCTOR CHAIN
// =======================================================
//
// Construction order: Base → Derived
// Destruction order: Derived → Base (reverse!)

class Base {
public:
  Base() { std::cout << "1️⃣  Base constructor\n"; }

  virtual ~Base() { std::cout << "3️⃣  Base destructor\n"; }
};

class Middle : public Base {
public:
  Middle() { std::cout << "2️⃣  Middle constructor\n"; }

  ~Middle() { std::cout << "2️⃣  Middle destructor\n"; }
};

class Derived : public Middle {
public:
  Derived() { std::cout << "3️⃣  Derived constructor\n"; }

  ~Derived() { std::cout << "1️⃣  Derived destructor\n"; }
};

//
// =======================================================
// 7. REAL-WORLD EXAMPLE: NOTIFICATION SYSTEM
// =======================================================
//

class Notification {
protected:
  std::string recipient;
  std::string message;
  std::string timestamp;

  std::string getCurrentTimestamp() const {
    time_t now = time(nullptr);
    char buf[20];
    strftime(buf, sizeof(buf), "%Y-%m-%d %H:%M:%S", localtime(&now));
    return std::string(buf);
  }

public:
  Notification(const std::string &r, const std::string &m)
      : recipient(r), message(m), timestamp(getCurrentTimestamp()) {}

  virtual ~Notification() = default;

  std::string formatHeader() const {
    return "[" + timestamp + "] To: " + recipient;
  }

  virtual void send() const {
    std::cout << formatHeader() << "\n";
    std::cout << "Message: " << message << "\n";
  }
};

class EmailNotification : public Notification {
private:
  std::string subject;

public:
  EmailNotification(const std::string &r, const std::string &m,
                    const std::string &s)
      : Notification(r, m), subject(s) {}

  void send() const override {
    std::cout << "\n📧 EMAIL NOTIFICATION\n";
    std::cout << formatHeader() << "\n";
    std::cout << "Subject: " << subject << "\n";
    std::cout << "Body: " << message << "\n";
    std::cout << "✅ Email delivered\n";
  }
};

class SMSNotification : public Notification {
private:
  std::string phoneNumber;
  static const int MAX_LENGTH = 160;

public:
  SMSNotification(const std::string &r, const std::string &m,
                  const std::string &phone)
      : Notification(r, m), phoneNumber(phone) {}

  void send() const override {
    std::cout << "\n📱 SMS NOTIFICATION\n";
    std::cout << formatHeader() << "\n";
    std::cout << "Phone: " << phoneNumber << "\n";

    std::string smsBody = message.length() > MAX_LENGTH
                              ? message.substr(0, MAX_LENGTH - 3) + "..."
                              : message;

    std::cout << "SMS: " << smsBody << "\n";
    std::cout << "✅ SMS sent (" << smsBody.length() << "/" << MAX_LENGTH
              << " chars)\n";
  }
};

class PushNotification : public Notification {
private:
  std::string deviceToken;
  std::string priority;

public:
  PushNotification(const std::string &r, const std::string &m,
                   const std::string &token, const std::string &prio)
      : Notification(r, m), deviceToken(token), priority(prio) {}

  void send() const override {
    std::cout << "\n🔔 PUSH NOTIFICATION\n";
    std::cout << formatHeader() << "\n";
    std::cout << "Device: " << deviceToken.substr(0, 8) << "...\n";
    std::cout << "Priority: " << priority << "\n";
    std::cout << "Alert: " << message << "\n";
    std::cout << "✅ Push delivered\n";
  }
};

//
// =======================================================
// 8. POLYMORPHISM THROUGH INHERITANCE
// =======================================================
//

void sendNotification(const Notification *notify) {
  notify->send(); // Polymorphic call!
}

//
// =======================================================
// 9. BEST PRACTICES
// =======================================================
//
// 1. Always use virtual destructors in base classes
// 2. Use 'override' keyword for clarity
// 3. Prefer composition over inheritance when possible
// 4. Keep inheritance hierarchies shallow (2-3 levels max)
// 5. Follow the "is-a" relationship test

//
// =======================================================
// 10. DEMONSTRATION
// =======================================================
//

int main() {
  std::cout << "=== Inheritance in C++ Demo ===\n\n";

  // ---- Basic Inheritance ----
  std::cout << "1. Basic Inheritance (Animal → Dog):\n";
  {
    Dog dog("Buddy", 3, "Golden Retriever");
    dog.makeSound();
    dog.sleep();
    dog.fetch();
    dog.displayBreed();
  }

  std::cout << "\n2. Multi-level Inheritance (Vehicle → Car → ElectricCar):\n";
  ElectricCar tesla("Tesla", 2024, 4, 100);
  tesla.start();
  tesla.charge();
  tesla.honk();

  // ---- Hierarchical Inheritance ----
  std::cout << "\n3. Hierarchical Inheritance (Shape → Circle, Rectangle):\n";
  Circle circle("Red", 5);
  Rectangle rect("Blue", 10, 20);

  circle.display();
  rect.display();

  // ---- Multiple Inheritance ----
  std::cout
      << "\n4. Multiple Inheritance (Duck: Animal + Flyable + Swimmable):\n";
  Duck duck("Donald", 2);
  duck.makeSound();
  duck.fly();
  duck.swim();

  // ---- Virtual Inheritance ----
  std::cout << "\n5. Virtual Inheritance (Diamond Problem Solution):\n";
  HybridDevice device(80);
  device.displayPower(); // Only ONE PowerSource instance!

  // ---- Constructor/Destructor Chain ----
  std::cout << "\n6. Constructor/Destructor Chain:\n";
  std::cout << "Creating Derived object:\n";
  {
    Derived d;
    std::cout << "Derived object exists here\n";
  }
  std::cout << "Derived object destroyed\n";

  // ---- Notification System ----
  std::cout << "\n7. Real-World Example (Notification System):\n";
  EmailNotification email("alice@example.com", "Your order has been shipped!",
                          "Order Update");
  email.send();

  SMSNotification sms("Bob", "Your verification code is 482910", "+1-555-0123");
  sms.send();

  PushNotification push("Charlie", "New message from Alice", "d8a3f4b2c1e5a9b7",
                        "high");
  push.send();

  // ---- Polymorphism ----
  std::cout << "\n8. Polymorphism Through Inheritance:\n";
  std::vector<Notification *> notifications = {&email, &sms, &push};

  for (Notification *notify : notifications) {
    sendNotification(notify);
  }

  // ---- Upcasting ----
  std::cout << "\n9. Upcasting (Derived → Base):\n";
  Dog myDog("Max", 5, "Labrador");
  Animal *animalPtr = &myDog; // Upcasting
  animalPtr->makeSound();     // Calls Dog's version (polymorphism!)

  std::cout << "\n=== Demo Complete ===\n";
  return 0;
}

/*
📘 Key Insights:

1. **Inheritance Types**:
   - Single: One parent
   - Multi-level: Chain of parents
   - Hierarchical: Multiple children
   - Multiple: Multiple parents (C++ only)

2. **Access Specifiers**:
   - Public inheritance: "is-a" relationship (most common)
   - Protected inheritance: Rare, "implemented-in-terms-of"
   - Private inheritance: Very rare, "implemented-using"

3. **Virtual Functions**:
   - Enable runtime polymorphism
   - Must use 'virtual' in base class
   - Use 'override' in derived class
   - Always virtual destructors!

4. **Diamond Problem**:
   - Occurs with multiple inheritance
   - Solution: Virtual inheritance
   - Prevents duplicate base class instances

5. **Constructor/Destructor Order**:
   - Construction: Base → Derived
   - Destruction: Derived → Base (reverse)

Best Practices:
- Use inheritance for "is-a" relationships
- Prefer composition for "has-a" relationships
- Keep hierarchies shallow (2-3 levels)
- Always virtual destructors in base classes
- Use 'override' keyword
- Follow Liskov Substitution Principle

When to Use Inheritance:
✅ Clear "is-a" relationship
✅ Shared behavior/data
✅ Code reuse within same family
✅ Polymorphism needed

When to Avoid:
❌ "has-a" relationship (use composition)
❌ Need runtime flexibility (use composition)
❌ Deep hierarchies
❌ Tight coupling concerns
*/